### ISSUE TITLE
Thread safety comprehensive examination

### ISSUE THREAD

**Issue by BenjaminTJohnson** [JCSDA/CRTMv3#111]
### Description

This issue is opened to address the need for exploring and testing the thread safety of the CRTM. Given the increasing use of multi-threaded environments in scientific computing, ensuring that CRTM is thread-safe is crucial for its reliability and performance in such contexts.

### Objectives

1. **Identify Potential Thread Safety Issues**: Review the CRTM codebase, particularly focusing on shared resources, global variables, and parallel execution segments, to identify potential thread safety issues.

2. **Develop Thread Safety Tests**: Create a suite of tests that can be used to assess the thread safety of CRTM. These tests should cover various scenarios of concurrent access and modifications by multiple threads.

3. **Implement Fixes for Identified Issues**: Based on the findings from the thread safety tests, implement necessary changes to ensure thread safety. This may involve modifying global variables, adjusting locks, or redesigning certain parts of the code.

4. **Documentation**: Update the documentation to reflect any changes made to the codebase and provide guidelines on how to use CRTM in multi-threaded environments.

5. **Continuous Integration Testing**: Integrate the thread safety tests into the continuous integration system to ensure ongoing thread safety compliance.

### Expected Outcomes

- A report on the current state of thread safety in CRTM.
- A robust set of thread safety tests that can be regularly used to validate the CRTM.
- An updated, thread-safe version of CRTM, if necessary.
- Documentation that guides users on how to safely run CRTM in multi-threaded environments.

### Call for Contributions

We encourage contributions from the community, especially from those who have experience in multi-threaded programming and thread safety. Contributions can be in the form of code reviews, writing tests, proposing fixes, or updating documentation.

### Contact

For more information or to contribute to this issue, please contact us.

**Comment by BenjaminTJohnson** [comment]
## Status update — thread-safety audit & fixes (via `feature/btj_REL-3.2.0`, also PR #277)

A round of work targeting objective (3) — the parallel radiative-transfer path — has landed. Summary of what's fixed, then a per-objective status and a closeability recommendation.

### Fixes landed

**1. `SAVE` / implicit-`SAVE` on runtime-populated coefficient arrays** (these made concurrent threads stomp shared scratch):
- `NESDIS_AMSU_SICEEM`, `NESDIS_AMSU_SnowEM`, `NESDIS_ATMS_SeaICE`, `NESDIS_ATMS_SnowEM`, `NESDIS_SSMI` — explicit `SAVE` attributes removed.
- `ODCAPS_AtmAbsorption`, `ODCAPS_Predictor` — inline `pointer => NULL()` (which imposes `SAVE` semantics) replaced with explicit `NULLIFY`.

**2. Channel-thread `!$OMP PARALLEL DO` loops in `CRTM_Forward_Module` / `CRTMTM_K_Matrix_Module`:**
- Shared `Error_Status` was being written inside the parallel region — a thread that hit `FAILURE` could be overwritten by a later `SUCCESS` from another thread, so the post-loop `IF (Error_Status == FAILURE) RETURN` could be silently skipped. Replaced with a per-thread `Err_Thread` + `REDUCTION(MAX:thread_error)`. Also caught a lowercase `Error_status =` typo inK-matrix that was aliasing the same shared result variable.
- Unindexed `RTV%...` / `RTV_Clear%...` flag broadcasts inside the loop → all per-thread elements written by every thread; changed to `RTV(nt)%...`.
- Off-by-one in the `n_inactive_channels` chunk-bucket mapping → out-of-bounds heap write `RTSolution(n_Channels+1, m)` when there are inactive boundary channels (e.g. the channel-subset cases).
- Missing clamp on the non-last `end_ch` → out-of-bounds `Process_Channel` read when `n_channel_threads * chunk_ch` overshoots.
- `AAvar(n_channel_threads)` was in the `PRIVATE` clause yet indexed `AAvar(nt)` — every thread allocated the whole array. Moved to SHARED, indexed by `nt` like its siblings (`RTV`, `AtmOptics`, `CSvar`, …).

**3. Per-channel state contamination in K-matrix:** the NLTE adjoint predictor and the Zeeman predictor are now reset at the top of each channel iteration. Previously they carried state across channels within a thread, so the `Atmosphere_K` Jacobians depended on ton and diverged from the serial reference (symptom: K-matrix tests pass at`OMP_NUM_THREADS` ∈ {1,2,4}, fail with "Atmosphere_K Jacobians are different!" at ≥ 8).                                                                                                   
**4. K-matrix channel-thread OpenMP re-enabled** behind a compiler-conditional gate — only legacy classic ifort (`__INTEL_COMPILER && !__INTEL_LLVM_COMPILER`) stays on the serial fallback (no local toolchain to validate it); gfortran 13.x and ifx 2026 verified clean.

**5. Runtime OpenMP controls:** `OMP_NUM_THREADS` honored at runtime, `OPENMP=OFF` build supported, empty `OMP_NUM_THREADS` treated the same as unset.
                                                                                                                                                                                          ### Validation
Full `ctest` (195 tests) passes at `OMP_NUM_THREADS` = 1, 4, and 8 (gfortran 13.3). The failure set is byte-identical with and without the thread-safety commits — i.e. they're correctness-neutral, as a race fix should be. The races that were found surfaced as SIGSEGV / heap corruption at `OMP_NUM_THREADS` ≥ 4 (Forward) and ≥ 19 (TL), and as Jacobian mismatches at ≥ 8 (K-matrix) — all now clean.

### Per-objective status
  | # | Objective | Status |
  |---|-----------|--------|
  | 1 | Identify vulnerabilities | **Mostly done** for the parallel RT path (the four entry points and their callees inside the channel loop). Not exhaustively audited: ODZeeman, module-level state in the remaining SfcOptics/AtmScatter modules, and coefficient I/O during `CRTM_Init`/`CRTM_Destroy` (single-threaded by contract — fine, but should be stated explicitly). |
  | 2 | Concurrent-access test procedures | **Partial.** `test_OMPoverChannels` + `test_OMP_Speedup` regression tests; whole suite validated across OMP thread counts. No dedicated race-detector run (ThreadSanitizer / Helgrind) or high-thread × multi-sensor stress test yet. |
  | 3 | Implement corrections | **Done** for everything found above. |
  | 4 | Multi-threaded usage docs | **Partial.** README has OpenMP thread-control notes (`OMP_NUM_THREADS`, `OPENMP=OFF`). Still missing a "threading model & guarantees" section: init/destroy are single-threaded; the RT entry points use internal channel-thread parallelism; statement on whether concurrent host-thread calls on shared coefficient structures are supported; legacy classic ifort → serial fallback. |
  | 5 | Thread-safety in CI | **Not started.** CI should run `ctest` at an elevated `OMP_NUM_THREADS` (and ideally a `-fsanitize=thread` build). |

**Comment by BenjaminTJohnson** [comment]
## Status update — thread-safety audit & fixes

Landing via `feature/btj_REL-3.2.0` (a related subset is in PR #277). Summary of fixes, the objective-1 audit, and a per-objective status / closeability call.

### Fixes landed
- **`SAVE` / implicit-`SAVE` on runtime-populated coefficient scratch:** `NESDIS_AMSU_SICEEM`, `NESDIS_AMSU_SnowEM`, `NESDIS_ATMS_SeaICE`, `NESDIS_ATMS_SnowEM`, `NESDIS_SSMI` — explicit `SAVE` removed; `ODCAPS_AtmAbsorption`, `ODCAPS_Predictor` — inline `pointer => NULL()` (which imposes `SAVE`) → explicit `NULLIFY`.
- **Channel-thread `!$OMP PARALLEL DO` loops in `CRTM_Forward_Module` / `CRTM_Tangent_Linear_Module` / `CRTM_K_Matrix_Module`:**
  - shared `Error_Status` written inside the parallel region (a `FAILURE` could be clobbered by a later `SUCCESS` from another thread, so the post-loop `IF (Error_Status==FAILURE) RETURN` was skippable) → per-thread `Err_Thread` + `REDUCTION(MAX:thread_error)`; also fixed a lowercase `Error_status =` typo in K-matrix that aliased the same shared result;
  - unindexed `RTV%…` / `RTV_Clear%…` flag broadcasts → `RTV(nt)%…`;
  - off-by-one in the `n_inactive_channels` chunk-bucket mapping → OOB heap write `RTSolution(n_Channels+1,m)` with inactive boundary channels;
  - missing clamp on the non-last `end_ch` → OOB `Process_Channel` read on overshoot;
  - `AAvar(n_channel_threads)` in the `PRIVATE` clause yet indexed `AAvar(nt)` → moved to SHARED like its siblings.
- **Per-channel state contamination in K-matrix:** NLTE adjoint predictor and Zeeman predictor now reset per channel (was carrying state across channels within a thread → Jacobians depended on the thread/channel partition; symptom: K-matrix tests pass at `OMP_NUM_THREADS` ∈ {1,2,4}, fail with "Atmosphere_K Jacobians are different!" at ≥ 8).
- **K-matrix channel-thread OpenMP re-enabled** behind a compiler-conditional gate — only legacy classic ifort stays on the serial fallback; gfortran 13.x and ifx 2026 verified clean.
- **Runtime OpenMP controls:** `OMP_NUM_THREADS` honored at runtime, `OPENMP=OFF` build supported, empty `OMP_NUM_THREADS` treated as unset.
- **Hygiene:** dropped the dead `Zeeman/Zeeman_Utility.f90` (an unused geomagn-level `SAVE`) from the `libcrtm` build — runtime ODZeeman uses `BeCoeff`.

### Validation
Full `ctest` (195 tests) passes at `OMP_NUM_THREADS` = 1, 4, 8 (gfortran 13.3). The failure set is byte-identical with and without the thread-safety commits — i.e. they're correctness-neutral, as race fixes should be. The races surfaced as SIGSEGV / heap corruption at `OMP_NUM_THREADS` ≥ 4 (Forward) and ≥ 19 (TL), and as Jacobian mismatches at ≥ 8  (K-matrix) — all now clean.

### Objective 1 audit — completed
Audited all source files compiled into `libcrtm` (offline tools — `TauProd/`, `TauRegress/`, `NLTE/Compute_*`, `*_ASC2NC`, `User_Code/` — out of scope). Findings beyond the fixes above:
- **No `COMMON` blocks; no lazy "first-call" init; no `=> NULL()`-implies-`SAVE`** pointer declarations in compiled non-`*_Define` procedures (remaining `=> NULL()` is commented-out code or legitimate derived-type-component defaults).
- **Only `!$OMP` regions in the library are the 4 RT entry points.** `CRTM_Adjoint_Module` parallelizes over **profiles** (`PRIVATE(m,Opt,AncillaryInput)`), aggregates errors via a per-element `ret(:)` array (the correct pattern — no shared `Error_Status` write), `profile_solution`'s state is all procedure-local, its host-associated accesses are `m`-indexed (distinct per thread) or read-only, and `Post_Process_RTSolution` is contained with explicit args — **verified thread-safe**.
- **Remaining `SAVE`s are all benign:** the `Coefficients/CRTM_*Coeff.f90` holders (`SC`, `TC`, `CloudC`, `AeroC`, `BeC`, `MWwaterC`, `PARMIOC`, `IR*C`, `VIS*C`, per-algorithm `ODAS/ODPS/ODSSU/ODZeeman_TauCoeff::TC`) are written only by `*_Load`/`*_Destroy` (single-threaded by API contract), read-only after Init — the standard design; `CRTM_Model_Profiles`'s `Level_Temperature`/`Level_Absorber` are compile-time `DATA` constants; `CRTM_Parameters::MAX_N_CHANNELS` is vestigial (its mutators are never called); `Zeeman_Utility::BField` is dead (now out of the build); `CRTM_Fastem3` / `CRTM_OceanEM_AMSRE` / `CRTM_surface_ratch but aren't compiled into `libcrtm`.
- **One contract to document (→ obj 4):** `CRTM_Init`/`CRTM_Destroy` are single-threaded and must not overlap RT calls or each other; after a successful `CRTM_Init`, the four RT entry points are safe to call (Forward/TL/K with internal channel-thread parallelism; Adjoint with internal profile parallelism).

### Per-objective status
  | # | Objective | Status |
  |---|-----------|--------|
  | 1 | Identify vulnerabilities | **Done** for the compiled library (see above) — the only true vulnerabilities found were already fixed; residual `SAVE`s are benign. |
  | 2 | Concurrent-access test procedures | **Partial.** `test_OMPoverChannels` + `test_OMP_Speedup` regression tests; whole suite validated at `OMP_NUM_THREADS` 1/4/8. No dedicated ThreadSanitizer/Helgrind build or high-thread × multi-sensor stress test yet. |
  | 3 | Implement corrections | **Done** for everything found. |
  | 4 | Multi-threaded usage docs | **Partial.** README has OpenMP thread-control notes; still missing a "threading model & guarantees" section (the Init/Destroy contract above; legacy classic ifort → serial fallback). |
  | 5 | Thread-safety in CI | **Not started.** CI should run `ctest` at an elevay also a `-fsanitize=thread` build. |

**Comment by BenjaminTJohnson** [comment]
## Thread-safety audit & fixes — complete (via `feature/btj_REL-3.2.0`; a subset is in PR #277)

All five objectives addressed.

### Fixes landed
- **`SAVE` / implicit-`SAVE` on runtime-populated coefficient scratch:** `NESDIS_AMSU_SICEEM`, `NESDIS_AMSU_SnowEM`, `NESDIS_ATMS_SeaICE`, `NESDIS_ATMS_SnowEM`, `NESDIS_SSMI` (explicit `SAVE` removed); `ODCAPS_AtmAbsorption`, `ODCAPS_Predictor` (inline `pointer => NULL()` → explicit `NULLIFY`).
- **Channel-thread `!$OMP PARALLEL DO` loops in `CRTM_Forward_Module` / `CRTM_Tangent_Linear_Module` / `CRTM_K_Matrix_Module`:** shared `Error_Status` write inside the parallel region (a `FAILURE` could be clobbered by a later `SUCCESS` from another thread, making the post-loop `IF (Error_Status==FAILURE) RETURN` skippable) → per-thread `Err_Thread` + `REDUCTION(MAX:thread_error)`, plus a lowercase `Error_status =` typo in K-matrix that aliased the same result; unindexed `RTV%…` / `RTV_Clear%…` flag broadcasts → `RTV(nt)%…`; off-by-one in the `n_inactive_channels` chunk-bucket mapping → OOB heap write `RTSolution(n_Channels+1,m)` with inactive boundary channels; missing clamp on the non-last `end_ch` → OOB `Process_Channel` read on overshoot; `AAvar(n_channel_threads)` in the `PRIVATE` clause yet indexed `AAvar(nt)` → SHARED.
- **Per-channel state contamination in K-matrix:** NLTE adjoint predictor and Zeeman predictor reset per channel (was carrying state across channels within a thread → `Atmosphere_K` Jacobians depended on the thread/channel partition; failed at `OMP_NUM_THREADS` ≥ 8).
- **K-matrix channel-thread OpenMP re-enabled** behind a compiler-conditional gate — only legacy classic `ifort` (pre-LLVM) uses the serial fallback; gfortran, Intel-LLVM (`ifx`) and nvfortran run the parallel path.
- **Runtime OpenMP controls:** `OMP_NUM_THREADS` honored at run time, `OPENMP=OFF` build supported, empty `OMP_NUM_THREADS` coerced to 1 (libgomp misbehaves on an empty value).
- **Hygiene:** dropped the dead `Zeeman/Zeeman_Utility.f90` (an unused geomagnetic-LUT loader with a module-level `SAVE`) from the `libcrtm` build.

### Objective-1 audit (compiled `libcrtm`)
No remaining vulnerabilities beyond the fixes above. No `COMMON` blocks; no lazy first-call init; remaining `=> NULL()` in compiled code is commented-out or legitimate derived-type-component defaults; the only `!$OMP` regions are the 4 RT entry points (`CRTM_Adjoint` parallelizes over **profiles** and aggregates errors via a per-element `ret(:)` array — correct; `profile_solution`'s state is procedure-local and its host-associated accesses are `m`-indexed or read-only — verified safe). Residual `SAVE`s are all benign: the `Coefficients/CRTM_*Coeff.f90` holders are written only by `*_Load`/`*_Destroy` (single-threaded by API contract), read-only after Init; `CRTM_Model_Profiles`'s `Level_*` arrays are compile-time `DATA` constants; `CRTM_Parameters::MAX_N_CHANNELS` is vestigial (never written); `CRTM_Fastem3` / `CRTM_OceanEM_AMSRE` / `CRTM_surface_ir_emissivity` have `SAVE` scratch but aren't compiled into `libcrtm`.

### Validation
Full `ctest` (198 tests) passes at `OMP_NUM_THREADS` = 1, 4, 8 (gfortran 13.3). The failure set is byte-identical with and without the thread-safety commits — the fixes are correctness-neutral. The races had surfaced as SIGSEGV/heap corruption at `OMP_NUM_THREADS` ≥ 4 (Forward), ≥ 19 (TL), and `Atmosphere_K` mismatches at ≥ 8 (K-matrix).

### Testing & docs
- New `test_OMP_Consistency` (in the `CRTM_Tests` label, so it runs in jedi-ci): runs `CRTM_Forward` + `CRTM_K_Matrix` at `OMP_NUM_THREADS=1` then a sweep (2, 4, 8, … up to the host core count) and asserts `RTSolution` / `RTSolution_K` / `Atmosphere_K` / `Surface_K` are **bit-identical** to the serial run (CRTM's `==`). Self-consistency — no reference files. No-op (PASS) under a non-OpenMP build or a single-thread host. Sensors: `atms_n21`, `abi_g18`, `cris399_npp`.
- README gained an "OpenMP and thread safety" section: the `-DOPENMP` build option, `OMP_NUM_THREADS` behaviour, which entry points parallelize over what (Forward/TL/K: channels ±profiles; Adjoint: profiles), the host-application contract (`CRTM_Init`/`CRTM_Destroy` single-threaded; the RT entry points are read-only after Init but must be called from a single host thread; non-overlapping array arguments), and the legacy-`ifort` serial f

### Objectives
1. Identify vulnerabilities — **done** (audit above). 
2. Concurrent-access tesistency`). 
3. Implement corrections — **done**. 
4. Multi-threaded usage docs —**done** (README). 
5. Thread-safety in CI — **done in effect** (`test_OMP_Consistency` runs in jedi-ci on every PR via the `CRTM_Tests` label; a dedicated high-thread CI lane could be added to the jedi-ci action separately).
